### PR TITLE
no WriteIntentError for inconsistent reads

### PR DESCRIPTION
**Pull request by tbg**
proposition for #1352, to remove the special-casing of `WriteIntentError`. This removes inconsistent reads complete from `WriteIntentError` handling.

I marked this as WIP since I'll likely brush some things up but it's certainly ready for a good look.
The important changes are in `mvccGetInternal` which has seen quite a bit of (as I find, clarifying) refactoring, `InternalRangeLookup`, Range and a little bit in Store.
- read commands return their skipped intents separately
- different async resolve path disjoint from WriteIntentError:
  proposing replica is in charge (eventually sends it up to the
  store via resolveWriteIntentError)
- refactored MVCC's internal get function quite a bit
- some test tweaks

### CODE REVIEW

File: proto/data.go [hunk 1]
```diff
@@ -511,8 +511,20 @@ func (t Transaction) Short() string {
...
 // IsInline returns true if the value is inlined in the metadata.
-func (mvcc *MVCCMetadata) IsInline() bool {
-	return mvcc.Value != nil
+func (meta MVCCMetadata) IsInline() bool {
+	return meta.Value != nil
+}
+
+// HasWriteIntentError returns whether the metadata has an open intent which
+// has not been laid down by the given transaction (which may be nil).
+func (meta *MVCCMetadata) HasWriteIntentError(txn *Transaction) bool {
```
Comment (tamird) [inline proto/data.go]: nit: inconsistent receiver type

**Comment by tschottdorf** [comment]
an alternative option could be extending the reply protos for those commands which can return skipped intents. Could be better given that there's almost never anything to return.

File: storage/engine/mvcc.go [hunk 2]
```diff
@@ -372,12 +373,13 @@ var getBufferPool = sync.Pool{
 // keyB : MVCCMetadata of keyB
 // ...
 //
-// The consistent parameter indicates how write intents should be
-// handled. In inconsistent mode, we return a valid value if one was
-// found in addition to the WriteIntentError.
-func MVCCGet(engine Engine, key proto.Key, timestamp proto.Timestamp, consistent bool, txn *proto.Transaction) (*proto.Value, error) {
+// The consistent parameter indicates that intents should cause
+// WriteIntentErrors. If set to false, intents are ignored; keys with
```
Comment (tamird) [inline storage/engine/mvcc.go]: the word "keys" here refers to the single argument "key", yes?

File: storage/engine/mvcc.go [hunk 3]
```diff
@@ -423,57 +425,64 @@ type getValueFunc func(engine Engine, start, end proto.EncodedKey,
 // The consistent parameter specifies whether reads should ignore any
 // pending write intents and read the most recent _committed_ value
 // instead. In the event that an inconsistent read does encounter
-// intents, the intent is returned via a WriteIntentError, in addition
-// to the result.
+// an intent (currently there can only be one), it is returned via a the
```
Comment (tamird) [inline storage/engine/mvcc.go]: via the

File: storage/engine/mvcc.go [hunk 4]
```diff
@@ -423,57 +425,64 @@ type getValueFunc func(engine Engine, start, end proto.EncodedKey,
...
+			return nil, nil, err
 		}
-		return meta.Value, nil
-	}
-	// If we're doing inconsistent reads and there's an intent, we
-	// ignore the intent by insisting that the timestamp we're reading
-	// at is a historical timestamp < the intent timestamp. However, we
-	// return a write intent error so that the intents can be resolved.
-	var wiErr error
-	if !consistent && meta.Txn != nil && !timestamp.Less(meta.Timestamp) {
```
Comment (tamird) [inline storage/engine/mvcc.go]: why reorder these?

File: storage/engine/mvcc.go [hunk 5]
```diff
@@ -423,57 +425,64 @@ type getValueFunc func(engine Engine, start, end proto.EncodedKey,
...
+	var err error
 	// First case: Our read timestamp is ahead of the latest write, or the
-	// latest write and current read are within the same transaction.
-	if !timestamp.Less(meta.Timestamp) ||
-		(meta.Txn != nil && txn != nil && bytes.Equal(meta.Txn.ID, txn.ID)) {
-		if meta.Txn != nil && (txn == nil || !bytes.Equal(meta.Txn.ID, txn.ID)) {
-			// Trying to read the last value, but it's another transaction's
-			// intent; the reader will have to act on this.
-			return nil, &proto.WriteIntentError{Intents: []proto.WriteIntentError_Intent{{Key: key, Txn: *meta.Txn}}}
+	// latest write and current read are within the same (non-nil) transaction.
```
Comment (tamird) [inline storage/engine/mvcc.go]: this comment no longer obviously describes the condition it's documenting

File: storage/engine/mvcc.go [hunk 6]
```diff
@@ -423,57 +425,64 @@ type getValueFunc func(engine Engine, start, end proto.EncodedKey,
...
+	if !timestamp.Less(meta.Timestamp) && meta.HasWriteIntentError(txn) {
+		// Trying to read the last value, but it's another transaction's
+		// intent; the reader will have to act on this.
+		return nil, nil, &proto.WriteIntentError{
+			Intents: []proto.Intent{{Key: key, Txn: *meta.Txn}},
 		}
+		// TODO remove
+	} else if ownIntent := meta.IsIntentOf(txn); !timestamp.Less(meta.Timestamp) || ownIntent {
+		// We want the latest value, and it's not an intent or our intent, both
```
Comment (tamird) [inline storage/engine/mvcc.go]: i'm having trouble understanding this comment

File: storage/engine/mvcc.go [hunk 7]
```diff
@@ -423,57 +425,64 @@ type getValueFunc func(engine Engine, start, end proto.EncodedKey,
...
 		// but it's got a different epoch. This can happen if the
 		// txn was restarted and an earlier iteration wrote the value
 		// we're now reading. In this case, we skip the intent.
-		if meta.Txn != nil && txn.Epoch != meta.Txn.Epoch {
+		if ownIntent && txn.Epoch != meta.Txn.Epoch {
 			valueKey, err = getValue(engine, latestKey.Next(), MVCCEncodeKey(key.Next()), value)
 		} else {
-			var ok bool
-			ok, _, _, err = engine.GetProto(latestKey, value)
+			ok, _, _, err := engine.GetProto(latestKey, value)
```
Comment (tamird) [inline storage/engine/mvcc.go]: are you intentionally shadowing `err` here?

File: storage/engine/mvcc.go [hunk 8]
```diff
@@ -423,57 +425,64 @@ type getValueFunc func(engine Engine, start, end proto.EncodedKey,
...
+		// TestMVCCReadWithPushedTimestamp).
 		latestKey := mvccEncodeTimestamp(metaKey, meta.Timestamp)
 
 		// Check for case where we're reading our own txn's intent
 		// but it's got a different epoch. This can happen if the
 		// txn was restarted and an earlier iteration wrote the value
 		// we're now reading. In this case, we skip the intent.
-		if meta.Txn != nil && txn.Epoch != meta.Txn.Epoch {
+		if ownIntent && txn.Epoch != meta.Txn.Epoch {
 			valueKey, err = getValue(engine, latestKey.Next(), MVCCEncodeKey(key.Next()), value)
```
Comment (tamird) [inline storage/engine/mvcc.go]: can you handle this error right here?

**Comment by bdarnell** [comment]
LGTM, although it's kind of a scary and intrusive change. It's unfortunate that we have to add so many `_` results everywhere, although it also highlights the need for this change if we have so many places that are incapable of handling WriteIntentErrors. All of those places have hard-coded `consistent=true`, though. One more option would be to have different functions for consistent and inconsistent modes instead of a boolean parameter, so the return types could be different. This would add a little complexity to the minority of cases that may perform inconsistent reads, but they already need to be able to handle the returned intents. I'm not sure it's worth it though since you'd still need the same changes to mvccGetInternal and churn at each call site to remove the consistent parameter.

File: proto/data.go (hunk 1 above)
Comment (tschottdorf) [inline proto/data.go]: done.

File: storage/engine/mvcc.go (hunk 2 above)
Comment (tschottdorf) [inline storage/engine/mvcc.go]: yes. Updated the comment.

File: storage/engine/mvcc.go (hunk 3 above)
Comment (tschottdorf) [inline storage/engine/mvcc.go]: done. Also updated the comment.

File: storage/engine/mvcc.go (hunk 4 above)
Comment (tschottdorf) [inline storage/engine/mvcc.go]: Likely due to intermediate changes no more present in the PR. I'm putting `!consistent` back to the front.

File: storage/engine/mvcc.go (hunk 5 above)
Comment (tschottdorf) [inline storage/engine/mvcc.go]: moved down where it belongs.

File: storage/engine/mvcc.go (hunk 6 above)
Comment (tschottdorf) [inline storage/engine/mvcc.go]: reworded it a little bit, let me know if it's clearer now.

File: storage/engine/mvcc.go (hunk 8 above)
Comment (tschottdorf) [inline storage/engine/mvcc.go]: I removed the function-scope err variable and added them in each of the cases. For this particular one it's now sharing one with the `else {...}` clause.

File: storage/engine/mvcc.go (hunk 7 above)
Comment (tschottdorf) [inline storage/engine/mvcc.go]: no. Changed with next commit.

**Comment by tschottdorf** [comment]
That's how I felt at first, but looking at the changes the majority of those `_` edits are in the tests.
Semantically, it's a tad weird to return those mostly nil intent slices around, but at the end of the day, we likely won't have two versions of `mvccGetInternal`, so as you said the churn mostly remains.
Introducing `MVCCGet{In,}Consistent` or something like that could be good sugar if we find that we have many of those across the codebase (which I'd argue we don't), but using them from the main execution path is kind of messy again (`executeCmd` needs to propagate intents upstream, so it would require a `executeCmdInconsistent` along with `MVCC{Scan,Get,InternalRangeLookup}Inconsistent`).

I think this current version is mergable, though I'm going to leave it open for discussion a bit longer.
PTAL if you haven't already.

File: proto/data.proto [hunk 9]
```diff
@@ -265,6 +265,13 @@ message Lease {
       (gogoproto.customname) = "RaftNodeID", (gogoproto.casttype) = "RaftNodeID"];
 }
 
+// Intent groups a key and a transaction.
+message Intent {
+  optional bytes key = 1 [(gogoproto.casttype) = "Key"];
+  optional Transaction txn = 2 [(gogoproto.nullable) = false];
+}
+
```
Comment (tamird) [inline proto/data.proto]: extra newline

File: storage/engine/mvcc.go [hunk 10]
```diff
@@ -420,56 +422,59 @@ type getValueFunc func(engine Engine, start, end proto.EncodedKey,
...
-			// intent; the reader will have to act on this.
-			return nil, &proto.WriteIntentError{Intents: []proto.WriteIntentError_Intent{{Key: key, Txn: *meta.Txn}}}
+	if !timestamp.Less(meta.Timestamp) && meta.HasWriteIntentError(txn) {
+		// Trying to read the last value, but it's another transaction's
+		// intent; the reader will have to act on this.
+		return nil, nil, &proto.WriteIntentError{
+			Intents: []proto.Intent{{Key: key, Txn: *meta.Txn}},
 		}
+	} else if ownIntent := meta.IsIntentOf(txn); !timestamp.Less(meta.Timestamp) || ownIntent {
+		// We are reading the latest value, which is not an intent or our intent
```
Comment (tamird) [inline storage/engine/mvcc.go]: "which is not an intent or our intent" is super confusing. what is it trying to say?

Comment (tschottdorf) [inline storage/engine/mvcc.go]: that whatever is there is either not an intent (in which case there's no intent, so no conflict) or it's the intent written by the same transaction (again, no conflict). How would you say it?

Comment (tamird) [inline storage/engine/mvcc.go]: "which is either an intent written by this transaction or not an intent at all (so there's no conflict)."

**Comment by tamird** [comment]
LGTM modulo one comment. please squash before merging

File: proto/data.proto (hunk 9 above)
Comment (tschottdorf) [inline proto/data.proto]: done.

File: storage/engine/mvcc.go (hunk 10 above)
Comment (tschottdorf) [inline storage/engine/mvcc.go]: done.

**Pull request merged**